translator/opencensus/oc_to_internal.go

```diff
@@ -0,0 +1,382 @@
+	}
+	return data.TraceState(sb.String())
+}
+
+func ocAttrsToInternal(ocAttrs *octrace.Span_Attributes) data.Attributes {
+	if ocAttrs == nil {
+		return data.Attributes{}
+	}
+
+	attrCount := len(ocAttrs.AttributeMap)
+	attrMap := make(data.AttributesMap, attrCount)
+	if attrCount > 0 {
+		values := data.NewAttributeValueSlice(attrCount)
+		i := 0
+		for key, ocAttr := range ocAttrs.AttributeMap {
+			switch attribValue := ocAttr.Value.(type) {
+			case *octrace.AttributeValue_StringValue:
+				values[i].SetString(attribValue.StringValue.GetValue())
+
+			case *octrace.AttributeValue_IntValue:
+				values[i].SetInt(attribValue.IntValue)
+
+			case *octrace.AttributeValue_BoolValue:
+				values[i].SetBool(attribValue.BoolValue)
+
+			case *octrace.AttributeValue_DoubleValue:
+				values[i].SetDouble(attribValue.DoubleValue)
+
+			default:
+				str := "<Unknown OpenCensus attribute value type>"
+				values[i].SetString(str)
+			}
+			attrMap[key] = values[i]
+			i++
+		}
+	}
+	droppedCount := uint32(ocAttrs.DroppedAttributesCount)
+	return data.NewAttributes(attrMap, droppedCount)
+}
+
+func ocSpanKindToInternal(ocKind octrace.Span_SpanKind, ocAttrs *octrace.Span_Attributes) data.SpanKind {
+	switch ocKind {
+	case octrace.Span_SERVER:
+		return data.SpanKindSERVER
+
+	case octrace.Span_CLIENT:
+		return data.SpanKindCLIENT
+
+	case octrace.Span_SPAN_KIND_UNSPECIFIED:
+		// Span kind field is unspecified, check if TagSpanKind attribute is set.
+		// This can happen if span kind had no equivalent in OC, so we could represent it in
+		// the SpanKind. In that case the span kind may be a special attribute TagSpanKind.
+		if ocAttrs != nil {
+			kindAttr := ocAttrs.AttributeMap[tracetranslator.TagSpanKind]
+			if kindAttr != nil {
+				strVal, ok := kindAttr.Value.(*octrace.AttributeValue_StringValue)
+				if ok && strVal != nil {
+					var otlpKind data.SpanKind
+					switch tracetranslator.OpenTracingSpanKind(strVal.StringValue.GetValue()) {
+					case tracetranslator.OpenTracingSpanKindConsumer:
+						otlpKind = data.SpanKindCONSUMER
+					case tracetranslator.OpenTracingSpanKindProducer:
+						otlpKind = data.SpanKindPRODUCER
+					default:
+						return data.SpanKindUNSPECIFIED
+					}
+					delete(ocAttrs.AttributeMap, tracetranslator.TagSpanKind)
+					return otlpKind
+				}
+			}
+		}
+		return data.SpanKindUNSPECIFIED
+
+	default:
+		return data.SpanKindUNSPECIFIED
+	}
+}
+
+func ocEventsToInternal(ocEvents *octrace.Span_TimeEvents) (ptrs []*data.SpanEvent, droppedCount uint32) {
+	if ocEvents == nil {
+		return
+	}
+
+	droppedCount = uint32(ocEvents.DroppedMessageEventsCount + ocEvents.DroppedAnnotationsCount)
+
+	eventCount := len(ocEvents.TimeEvent)
+	if eventCount == 0 {
+		return
+	}
+
+	// Create slice that holds pointers and another slice that holds structs at once
+	// to avoid individual allocations of structs.
+	ptrs = make([]*data.SpanEvent, 0, eventCount)
+	content := data.NewSpanEventSlice(eventCount)
```

Comment:
Same as for NewSpanSplice, should this return directly the pointer?

Reply:
Replied above, same here.

If we return slice of pointers we will have to write code to remove elements from the slice if the source slice contains nil pointers. It will make the code more complex.

Reply:
There may be cases where we know exactly how many items we are going to add. Also here it will be not that bad because we will just return a sub-slice. Not pushing for this idea, but trying to understand if maybe we should have NewSpanEventSlice and NewSpanEventSlicePtrs in the data as helpers :)